Adaptors/Amqp/src/PullQueueStorage.cs

```diff
@@ -123,9 +112,57 @@ await ConnectionAmqp.Init(cancellationToken)
       for (var i = receivers_.Length - 1; i >= 0; --i)
       {
         cancellationToken.ThrowIfCancellationRequested();
-        var receiver = await receivers_[i];
-        var message = await receiver.ReceiveAsync(TimeSpan.FromMilliseconds(100))
+        Message? message  = null;
+        var      receiver = await receivers_[i];
+
+        for (var retry = 0; retry < Options.MaxRetries; i++)
+        {
+          try
+          {
+            message = await receiver.ReceiveAsync(TimeSpan.FromMilliseconds(100))
                                     .ConfigureAwait(false);
+            break;
+          }
+          catch (Exception e)
+          {
+            if (retry < Options.MaxRetries - 1)
+            {
+              try
+              {
+                await Task.Delay(retry * retry * baseDelay_,
+                                 cancellationToken)
+                          .ConfigureAwait(false);
+
+                var session = new Session(ConnectionAmqp.Connection);
+                receivers_[i] = CreateReceiver(session,
```

Comment:
If you create the session in CreateReceiver and await the receiver within the first try, it would avoid the need for two nested tries altogether.

Reply:
Errors happen during Session creation so you still need the other try catch